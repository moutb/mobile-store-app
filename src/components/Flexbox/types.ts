import React from 'react';

export type StyleProps = {
    column?: boolean;
    wrap?: boolean;
    align?: React.CSSProperties['alignItems'];
    justify?: React.CSSProperties['justifyContent'];
    gap?: string | number;
};

export type FlexboxProps<T extends keyof React.JSX.IntrinsicElements> =
    StyleProps & {
        children?: React.ReactNode;
        xs?: StyleProps;
        sm?: StyleProps;
        md?: StyleProps;
        lg?: StyleProps;
        xl?: StyleProps;
        as?: T;
    } & React.JSX.IntrinsicElements[T];

export type FlexboxComponent = <
    T extends keyof React.JSX.IntrinsicElements = 'div',
>(
    props: FlexboxProps<T>,
) => React.JSX.Element;
