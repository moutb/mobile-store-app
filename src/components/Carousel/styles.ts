import styled from 'styled-components';

export const Wrapper = styled.div`
    display: flex;
    align-items: center;
    flex-flow: column nowrap;
    gap: 0.5rem;
    width: 100%;
`;

export const NavButton = styled.button`
    background: none;
    border: none;
    font-size: var(--font-size-md);
    cursor: pointer;
    color: var(--color-primary);
    padding: 0 0.5rem;

    &:disabled {
        opacity: 0.4;
        cursor: not-allowed;
    }

    &:focus {
        outline: 2px solid currentColor;
    }
`;

export const CarouselTrack = styled.div`
    display: flex;
    overflow-x: hidden;
    scroll-behavior: smooth;
    scroll-snap-type: x mandatory;
    flex: 1;
    width: 100%;
    user-select: none;
`;

export const CarouselItem = styled.div<{ $visibleCount: number }>`
    flex: 0 0 calc(100% / ${({ $visibleCount }) => $visibleCount});
    scroll-snap-align: start;
`;

export const ProgressBar = styled.div`
    width: 100%;
    height: 1px;
    background-color: var(--color-secondary);
    margin-top: var(--spacing-6);
    position: relative;
    border-radius: 2px;
    cursor: pointer;
    user-select: none;
`;

export const ScrollThumb = styled.div`
    position: absolute;
    height: 100%;
    background-color: var(--color-primary);
    border-radius: 2px;
    transition: left 0.2s ease;
`;
