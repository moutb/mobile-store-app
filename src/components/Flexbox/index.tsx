import styled, { css } from 'styled-components';
import { FlexboxComponent, FlexboxProps, StyleProps } from './types';
import baseShouldForwardProp from '@styled-system/should-forward-prop';
import { media } from '@/utils/media';

const getFlexStyles = (props: StyleProps) => css`
    flex-direction: ${props.column ? 'column' : 'row'};
    flex-wrap: ${props.wrap ? 'wrap' : 'nowrap'};
    align-items: ${props.align || 'center'};
    justify-content: ${props.justify || 'space-between'};
    gap: ${typeof props.gap === 'number' ? `${props.gap}px` : props.gap || '0'};
`;

// improvement for avoiding props on DOM element
const internalProps = [
    'column',
    'wrap',
    'align',
    'justify',
    'gap',
    'xs',
    'sm',
    'md',
    'lg',
    'xl',
];

const shouldForwardProp = (prop: string) =>
    baseShouldForwardProp(prop) && !internalProps.includes(prop);

const StyledFlex = styled('div').withConfig({ shouldForwardProp })<
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    FlexboxProps<any>
>`
    display: flex;
    ${(props) => getFlexStyles(props)}

    ${(props) =>
        props.xs &&
        media('xs')(css`
            ${getFlexStyles(props.xs)}
        `)}
    ${(props) =>
        props.sm &&
        media('sm')(css`
            ${getFlexStyles(props.sm)}
        `)}
    ${(props) =>
        props.md &&
        media('md')(css`
            ${getFlexStyles(props.md)}
        `)}
    ${(props) =>
        props.lg &&
        media('lg')(css`
            ${getFlexStyles(props.lg)}
        `)}
    ${(props) =>
        props.xl &&
        media('xl')(css`
            ${getFlexStyles(props.xl)}
        `)}
`;

const Flexbox: FlexboxComponent = (props) => {
    const { as, children, ...rest } = props;
    return (
        <StyledFlex as={as} {...rest}>
            {children}
        </StyledFlex>
    );
};

export default Flexbox;
