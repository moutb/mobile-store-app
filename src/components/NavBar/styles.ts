import Image from 'next/image';
import styled from 'styled-components';

export const Nav = styled.nav`
    background-color: ${({ theme }) => theme.colors.background};
    padding: var(--spacing-4) 0;
`;

export const BackButton = styled.button`
    display: inline-flex;
    align-items: center;
    gap: var(--spacing-1);
    background: none;
    border: none;
    font-size: var(--font-size-sm);
    line-height: var(--font-size-sm);
    font-weight: var(--font-weight-thin);
    color: var(--color-text);
    cursor: pointer;
    text-transform: uppercase;

    &:focus {
        outline: 2px solid var(--color-primary);
        outline-offset: 2px;
    }
`;

export const Icon = styled(Image)`
    flex-shrink: 0;
`;
