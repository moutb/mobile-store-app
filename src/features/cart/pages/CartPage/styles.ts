'use client';

import styled from 'styled-components';

export const CartContainer = styled.section`
    /* 100vh - header */
    height: calc(100vh - var(--header-height));
    display: flex;
    flex-direction: column;
    padding-top: var(--spacing-6);
    gap: var(spacing-4);
    justify-content: stretch;
`;

export const CartTitle = styled.h1`
    font-size: var(--font-size-lg);
    font-weight: var(--font-weight-thin);
`;

export const EmptyMessage = styled.p`
    font-size: var(--font-size-base);
    text-align: center;
    margin-top: var(--spacing-6);
`;

export const CartItemsContainer = styled.div`
    display: flex;
    margin-top: var(--spacing-6);
    width: 100%;
    justify-content: space-between;
    padding-bottom: var(--spacing-6);
    flex: 1;
    flex-flow: column nowrap;

    & > * {
        border-bottom: 0.5px solid var(--color-primary);
    }
    & > *:nth-last-child(2),
    & > *:last-child {
        border-bottom: none;
    }
`;
