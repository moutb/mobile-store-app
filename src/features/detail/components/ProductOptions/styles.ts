import styled from 'styled-components';

export const OptionsContainer = styled.fieldset`
    border: none;
    width: 100%;
`;

export const OptionTitle = styled.legend`
    text-transform: uppercase;
    font-size: var(--product-option-title-font-size);
    font-weight: var(--product-option-title-font-weight);
    line-height: var(--product-option-title-line-height);
    margin-bottom: var(--product-option-title-margin-bottom);
    letter-spacing: var(--product-option-title-letter-spacing);
`;

export const ColorOptions = styled.div`
    display: flex;
    gap: var(--product-color-options-gap);
    margin-bottom: 1rem;
`;

export const ColorBox = styled.div<{ $selected: boolean }>`
    width: var(--product-color-options-width);
    height: var(--product-color-options-height);
    transition: border 0.2s ease-in-out;
    border-width: ${({ $selected }) => ($selected ? '2px' : '1px')};
    border-style: solid;
    border-color: ${({ theme, $selected }) =>
        $selected ? theme.colors.primary : theme.colors.light};
    box-shadow: inset 0 0 0 1px white;
    cursor: pointer;

    &:focus-visible {
        outline: 2px var(--color-focus);
        outline-offset: 2px;
        border-radius: 4px;
    }
`;

export const ColorName = styled.span`
    font-size: var(--product-color-option-name-font-size);
    font-weight: var(--product-color-option-name-font-weight);
`;

export const StorageGroup = styled.div`
    display: flex;
    flex-wrap: wrap;
`;

export const StorageOption = styled.div<{ $selected: boolean }>`
    padding: var(--product-storage-options-padding);
    transition: border 0.6s ease-in-out;
    border: 1px solid
        ${({ theme, $selected }) =>
            $selected ? theme.colors.primary : theme.colors.light};
    background: ${({ $selected }) => ($selected ? '#fff' : 'transparent')};
    cursor: pointer;
    font: inherit;
    font-weight: var(--product-storage-options-font-weight);

    &:focus-visible {
        outline: 2px var(--color-focus);
        outline-offset: 2px;
        border-radius: 4px;
    }
`;
