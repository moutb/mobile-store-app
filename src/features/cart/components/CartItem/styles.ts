import { media } from '@/styles/media';
import Image from 'next/image';
import styled from 'styled-components';
import { css } from 'styled-components';

export const ItemContainer = styled.div`
    display: flex;
    gap: var(--spacing-4);
    align-items: center;
    padding-bottom: var(--spacing-4);
`;

export const ProductImageWrapper = styled.div`
    position: relative;
    height: var(--product-image-wrapper-height);
    width: 40%;

    ${media('md')(css`
        height: var(--product-image-wrapper-md-width);
        width: var(--product-image-wrapper-md-width);
    `)}
`;

export const ProductImage = styled(Image)`
    flex-shrink: 0;
`;

export const Info = styled.div`
    flex: 1;
    display: flex;
    flex-direction: column;
    gap: var(--spacing-2);
    height: 100%;
    max-height: var(--product-info-max-height);
`;

const ProductTextInfo = css`
    font-size: var(--font-size-sm);
    font-weight: var(--font-weight-thin);
    text-transform: uppercase;
`;

export const Name = styled.h2`
    ${ProductTextInfo}
`;

export const Description = styled.p`
    ${ProductTextInfo}
`;

export const Price = styled.p`
    ${ProductTextInfo}
    margin-top: var(--spacing-4);
`;

export const Spacer = styled.div`
    flex-grow: 1;
    flex-shrink: 0;
    width: 1px;
`;

export const RemoveButton = styled.button`
    align-self: flex-start;
    background: none;
    border: none;
    padding: 0;
    color: var(--color-danger);
    font-size: var(--font-size-sm);
    cursor: pointer;

    &:hover {
        text-decoration: underline;
    }

    &:focus-visible {
        outline: 2px solid var(--color-primary);
        outline-offset: 2px;
    }
`;
