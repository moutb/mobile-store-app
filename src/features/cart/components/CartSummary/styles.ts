// components/cart/styles/CartSummary.styles.ts
import { media } from '@/styles/media';
import Link from 'next/link';
import styled from 'styled-components';
import { css } from 'styled-components';

export const SummaryContainer = styled.div`
    display: flex;
    flex-direction: column;
    gap: var(--spacing-4);
    margin-top: var(--spacing-6);
    padding-bottom: var(--spacing-2);
`;

export const Actions = styled.div`
    display: flex;
    gap: var(--spacing-2);
    flex-wrap: wrap;
    width: 100%;
    justify-content: space-between;
    align-items: center;

    ${media('md')(css`
        flex-wrap: nowrap;
        justify-content: flex-end;
        gap: var(--spacing-6);
    `)}
`;

export const TotalRow = styled.div`
    display: flex;
    justify-content: space-between;
    width: 100%;
    gap: var(--spacing-4);
    font-size: var(--font-size-sm);
    font-weight: var(--font-weight-thin);

    ${media('md')(css`
        justify-content: flex-end;
    `)}
`;

export const TotalLabel = styled.span`
    color: var(--color-text-secondary);
`;

export const TotalPrice = styled.span`
    color: var(--color-text);
`;

const ActionButton = css`
    padding: var(--spacing-4) 0;
    cursor: pointer;
    font-size: var(--font-size-xs);
    line-height: var(--font-size-xs);
    font-weight: var(--font-weight-thin);
    text-transform: uppercase;
    width: calc(50% - var(--spacing-2));
    white-space: nowrap;
    text-align: center;
    margin-top: var(--spacing-4);

    ${media('md')(css`
        max-width: 260px;
        padding: var(--spacing-4) var(--spacing-6);
        font-size: var(--font-size-sm);
        line-height: var(--font-size-sm);
        margin-top: 0;
    `)}
`;

export const OutlinedButton = styled(Link)`
    ${ActionButton}
    background: none;
    color: var(--color-text);
    border-width: 0.5px;
    border-style: solid;
    border-color: var(--color-primary);

    &:hover {
        background: var(--color-secondary);
    }

    &:focus-visible {
        outline: 2px solid var(--color-primary);
        outline-offset: 2px;
    }

    ${media('md')(css`
        order: -1;
    `)}
`;

export const FilledButton = styled.button`
    ${ActionButton}
    background: var(--color-primary);
    color: var(--color-white);
    border: none;

    &:hover {
        opacity: 0.9;
    }

    &:focus-visible {
        outline: 2px solid var(--color-primary);
        outline-offset: 2px;
    }
`;
