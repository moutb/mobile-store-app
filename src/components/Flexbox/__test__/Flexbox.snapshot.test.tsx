import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '@/styles/theme';
import Flexbox from '@/components/Flexbox';

describe('Flexbox snapshot', () => {
    it('matches snapshot', () => {
        const { container } = render(
            <ThemeProvider theme={theme}>
                <Flexbox column align="center" justify="center" gap={16}>
                    <div>Child 1</div>
                    <div>Child 2</div>
                </Flexbox>
            </ThemeProvider>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });
});
