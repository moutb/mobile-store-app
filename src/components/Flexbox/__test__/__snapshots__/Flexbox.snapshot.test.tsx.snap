// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flexbox snapshot matches snapshot 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: center;
  gap: 16px;
}

<div
  class="c0"
>
  <div>
    Child 1
  </div>
  <div>
    Child 2
  </div>
</div>
`;
